### DISCUSSION TITLE
OpenID Connect Redirect URI Regex

### DISCUSSION THREAD

**Discussion by hunoz** [authelia/authelia#4659] [Ideas]
Hello,

I wasn't sure if an issue should be opened regarding this or if we should keep it an idea, but is there a plan to implement regex for OIDC redirect URIs? This feature is currently implemented in many other OIDC applications such as Auth0, Authentik (though it does not work yet), etc. I do see that the standard says the redirect URI must be absolute (seen [here](https://www.rfc-editor.org/rfc/rfc6749#section-3.1.2)), was this the reason why this feature was not implemented?

**Comment by james-d-elliott** [comment]
I don't really see a compelling reason to add this, what the other apps do isn't going to dictate our decisions, especially when they're potentially very problematic decisions for security. Also doing some searching it doesn't look like Auth0 does support regex matching for this as far as I can see.

As to why we have not it's not compliant with the relevant specifications for both the [Authorization Code Flow](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) and [Implicit Flow](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthRequest), registered redirect_uri's must be compared with simple string comparison. From the spec:

`This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [[RFC3986]](https://openid.net/specs/openid-connect-core-1_0.html#RFC3986) (Simple String Comparison).`

As to why these other apps are blatantly ignoring it I'm not sure.

**Comment by james-d-elliott** [comment]
The OAuth 2.0 Threat Model also talks about this: https://tools.ietf.org/html/rfc6819#section-5.2.3.5

**Comment by hunoz** [comment]
I do completely agree that the direction of one application should not influence another. This question was mainly to see if it was omitted for security reasons (as mentioned above) or if it was a feature that was perhaps overlooked. Thank you for the clarification!

**Reply by james-d-elliott** [reply]
I think if we were given a reasonable use case we may in the future, but generally speaking dynamically registered (yet to be implemented) and static URI's are entirely sufficient for all use cases. 

If we were to implement it, it would be very unlikely to be actual regex and instead we'd support use of a `*` which would represent any URI valid character that is in the range `[a-zA-Z0-9-]`. i.e. `https://*.example.com/*` would match `https://abc.example.com/callback` but not `https://abc.sub.example.com/callback` or `https://abc.example.com/api/callback`.

**Reply by hunoz** [reply]
I wonder if maybe locking it into the domains setting would also be good, possibly a good middle ground for better security than allowing putting in any domain with the asterisk. Maybe a boolean setting like `allow_domain_redirect_uris` that trusts domains and subdomains specified from the https://www.authelia.com/configuration/session/introduction/#domain setting?

**Comment by james-d-elliott** [comment]
So an additional discussion in the chat lead me to link all of the relevant information regarding this. We at this stage have no plans to match Redirect URI's that are not explicitly configured or registered with the exclusion of a special rule that applies to loopback addresses which we already abide by. This is going to be a very lengthy post with lots of technical documents linked, I apologize but I think it's important to post this so hopefully some other implementers (or users who are interested in this feature) either respond and show some clear contradiction or they remedy their non-compliant implementations. 

The [OpenID Connect 1.0 - Authorization Code Flow - Authentication Request (Section 3.1.2.1)](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) clearly outlines the rules for matching redirect URI's:

>REQUIRED. Redirection URI to which the response will be sent. This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [[RFC3986]](https://openid.net/specs/openid-connect-core-1_0.html#RFC3986) (Simple String Comparison). When using this flow, the Redirection URI SHOULD use the https scheme; however, it MAY use the http scheme, provided that the Client Type is confidential, as defined in Section 2.1 of OAuth 2.0, and provided the OP allows the use of http Redirection URIs in this case. The Redirection URI MAY use an alternate scheme, such as one that is intended to identify a callback into a native application.

Where [RFC3986 Section 6.2.1](https://www.rfc-editor.org/rfc/rfc3986#section-6.2.1) and all of section 6 clearly lacks any kind of mention of wildcards; in fact there is not a single term in that RFC that mentions wildcards. I think based on my research people have clearly misunderstood the use of the `*` in this RFC which has no specific guidance. I think it is unsafe to assume this means the string comparison should be done with wildcards or regex. There is in my opinion significant evidence to suggest this should not be treated in this way and is instead being used as a way to communicate the way you can check a URL is actually a valid URL in a general sense, not specifically for comparing two separate URLs. 

In addition the [RFC3986](https://www.rfc-editor.org/rfc/rfc3986) actually has a very clear clue as to what the `*` character represents in my opinion. When you look at them describing the port section of a URI you can see this clear clue for yourself if you analyze it closely in [Appendix A](https://www.rfc-editor.org/rfc/rfc3986#appendix-A). The text reads ` port = *DIGIT`. Now a  few questions here. Does `*` mean a wildcard? Can any character reasonably be assumed to be valid for a port? That seems absurd to me. What if they meant that a port can have more than one digit or be absent? That certainly matches with what we consider to be valid in the port area of a URI. To me this is a clear indication that they intended this character to represent a *similar* character as what exists in regular expressions for the purposes of describing a valid URI, not  for the purposes of how it's to be used with [OpenID Connect 1.0](https://openid.net/specs/openid-connect-core-1_0.html) / [OAuth 2.0](https://www.rfc-editor.org/rfc/rfc6749). It's very misguided to read it as if those sections should apply in that way to those unrelated specifications anyway. [OpenID Connect 1.0](https://openid.net/specs/openid-connect-core-1_0.html) *references* only *a single section* of [RFC3986](https://www.rfc-editor.org/rfc/rfc3986), not the entire document. As a means on how the `redirect_uri` parameter *should be compared*. 

When looking at [OAuth 2.0](https://www.rfc-editor.org/rfc/rfc6749) there are a few sections which discuss this topic. For example [OAuth 2.0 Section 3.1.2](https://www.rfc-editor.org/rfc/rfc6749#section-3.1.2) discusses that redirect URI's must be absolute URI's as per [RFC3986 Section 4.3](https://www.rfc-editor.org/rfc/rfc3986#section-4.3). In [OAuth 2.0 Section 5.2.3.5](https://www.rfc-editor.org/rfc/rfc6819#section-5.2.3.5) it discusses the security implications and the fact the registered URI's should be "FULL" URI's.

There is to my knowledge a single notable exception for loopback address (i.e. `127.0.0.1` and  `[::1]`) when used over the http scheme. Specifically that the port should be ignored for native applications. This comes from [RFC8252 Section 7.3](https://www.rfc-editor.org/rfc/rfc8252#section-7.3). Authelia already as discussed above handles this scenario. Interestingly this specific exception also is an indication that wildcards or regex patterns are not permitted as part of a providers URL comparison methodology. Why? Because if it was intended by the authors that providers could use wildcards or regex for matching then this specific exception would not need to exist at all.

Also if you look at the major differences coming in [OAuth 2.1](https://oauth.net/2.1/) there is a clear notice that the redirect URI's MUST be compared similar to the  [OpenID Connect 1.0](https://openid.net/specs/openid-connect-core-1_0.html) specification; using exact string matching.

Some implementations seem to arbitrarily assume all redirect URI's are regex strings, which has some particularly severe implications. For example if you assume the redirect_uri `https://myapp.example.com/oauth2/callback` registered at the provider as allowed is a regex string, then attacker x could register the domain `myapp1example.com`. Now if the client configured to handle `https://myapp.example.com/oauth2/callback` only allows the Implicit Flow or the Public Client Type, they can very easily phish Access Tokens, ID Tokens, and potentially Refresh Tokens from unsuspecting users. because the `.` in the domain means ANY string. So long as the string is valid for a domain, they can register it.

Some implementations have decided to make it an option an administrator has to consciously enable, and the wildcard redirect URI's (NOT regex), are specially configured separately to the standard compliant redirect URI's. This is a much more reasonable approach, much less unpalatable, and more likely the route we'd go down if we found reasonable legitimate use cases.

**Reply by danfate** [reply]
Hi:
I'm trying to get netbird with authelia,but never worked.

https://netbird.io/docs/integrations/identity-providers/self-hosted/using-netbird-with-keycloak

This app seems need the redirect_uris support regex or at least the wildcard.

**Reply by james-d-elliott** [reply]
What gives you the impression it needs this? There's nothing in their docs as far as I can see which claims this or provides reasonable justification for requiring it.

**Reply by danfate** [reply]
step 5. Valid redirect URIs: https://YOUR DOMAIN/`*`
<img width="788" alt="image" src="https://user-images.githubusercontent.com/8059612/215752548-d8bfbb92-2193-4e79-b720-0585a2a568f6.png">

**Reply by danfate** [reply]
And there's a pending pull request about using  Authentik as IDP for netbird, which said it should choose `Redirect URIs/Origins (RegEx)`:
https://github.com/netbirdio/docs/pull/23/files
<img width="726" alt="image" src="https://user-images.githubusercontent.com/8059612/215753229-1c9c4044-7bcf-41e5-a759-724ac4d611c5.png">

**Reply by james-d-elliott** [reply]
That's a configuration example not a stipulation that it requires that non-compliant and insecure feature. The best course of action is to try it and see which redirect URI's it actually requests and/or to ask them to provide clarification in an issue / discussion and to publish this to their official documentation regarding there actually being a requirement or not and what the actual expected redirect URIs are (or any related configuration) rather than people just being lazy when writing these docs as well as justification (provided they have the requirement) for their decision to ignore the specification.

**Reply by jrtashjian** [reply]
To fix the redirect issue, simply add `NETBIRD_AUTH_REDIRECT_URI="/peers"` to the `setup.env` file in NetBird and execute `./configure.sh`. Then set `https://YOUR DOMAIN/auth` as the redirect_uri in Authelia.

**Reply by james-d-elliott** [reply]
Thanks for including an answer.
